cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -0,0 +1,739 @@
+
+#include "arrow/compute/exec/expression_internal.h"
+#include "arrow/dataset/file_base.h"
+#include "arrow/dataset/scanner.h"
+#include "arrow/engine/substrait/extension_types.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/matchers.h"
+#include "arrow/util/key_value_metadata.h"
+
+using testing::ElementsAre;
+using testing::Eq;
+using testing::HasSubstr;
+
+namespace arrow {
+
+using internal::checked_cast;
+
+namespace engine {
+
+const std::shared_ptr<Schema> kBoringSchema = schema({
+    field("bool", boolean()),
+    field("i8", int8()),
+    field("i32", int32()),
+    field("i32_req", int32(), /*nullable=*/false),
+    field("u32", uint32()),
+    field("i64", int64()),
+    field("f32", float32()),
+    field("f32_req", float32(), /*nullable=*/false),
+    field("f64", float64()),
+    field("date64", date64()),
+    field("str", utf8()),
+    field("list_i32", list(int32())),
+    field("struct", struct_({
+                        field("i32", int32()),
+                        field("str", utf8()),
+                        field("struct_i32_str",
+                              struct_({field("i32", int32()), field("str", utf8())})),
+                    })),
+    field("list_struct", list(struct_({
+                             field("i32", int32()),
+                             field("str", utf8()),
+                             field("struct_i32_str", struct_({field("i32", int32()),
+                                                              field("str", utf8())})),
+                         }))),
+    field("dict_str", dictionary(int32(), utf8())),
+    field("dict_i32", dictionary(int32(), int32())),
+    field("ts_ns", timestamp(TimeUnit::NANO)),
+});
+
+std::shared_ptr<DataType> StripFieldNames(std::shared_ptr<DataType> type) {
+  if (type->id() == Type::STRUCT) {
+    FieldVector fields(type->num_fields());
+    for (int i = 0; i < type->num_fields(); ++i) {
+      fields[i] = type->field(i)->WithName("");
+    }
+    return struct_(std::move(fields));
+  }
+
+  if (type->id() == Type::LIST) {
+    return list(type->field(0)->WithName(""));
+  }
+
+  return type;
+}
+
+// map to an index-only field reference
+inline FieldRef BoringRef(FieldRef ref) {
+  auto path = *ref.FindOne(*kBoringSchema);
+  return {std::move(path)};
+}
+
+inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
+  if (expr.literal()) return expr;
+
+  if (auto ref = expr.field_ref()) {
+    return compute::field_ref(*ref->FindOne(*kBoringSchema));
+  }
+
+  auto modified_call = *CallNotNull(expr);
+  for (auto& arg : modified_call.arguments) {
+    arg = UseBoringRefs(arg);
+  }
+  return compute::Expression{std::move(modified_call)};
+}
+
+std::shared_ptr<Buffer> SubstraitFromJSON(util::string_view type_name,
+                                          util::string_view json) {
+  auto maybe_buf = internal::SubstraitFromJSON(type_name, json);
+  DCHECK_OK(maybe_buf.status());
+  return maybe_buf.ValueOrDie();
+}
```

Comment:
Maybe I'm missing something but do we really need this?  Seems like a call...
```
std::shared_ptr<Buffer> foo = SubstraitFromJSON(type_name, json);
```
...is no different than...
```
ASSERT_OK_AND_ASSIGN(std::shared_ptr<Buffer> foo, internal::SubstraitFromJSON(type_name, json));
```

Reply:
I'm guessing Ben wrote it this way to show that failure of `internal::SubstraitFromJSON()` is a problem with the tests (i.e. invalid JSON in the test case) rather than with the code that's actually under test. I guess it also saves a tiny amount of typing (it *is* used a lot). But I can change it if you want.

Reply:
Let's go ahead and change it.  It harms readability because future users will waste time trying to figure out the purpose of the function.

Also, I'm not really sure ValueOrDie accomplishes the goal you're describing.  An abort will crash the test in a way that logs may end up missing in CI.

Reply:
2156716